### PR TITLE
fix: attempt to only allow one deno process to update the node_modules folder at a time

### LINKED ISSUE #18039: Improve synchronization of multiple deno processes caching node_modules folder
We should add some basic synchronization around multiple processes creating the node_modules folder at the same time. Basically do whatever `cargo build` does and have each process wait on the other.

This is necessary because it seems like multiple processes creating junctions at the same time causes some weird issues with the junctions becoming invalid. Also, it's duplicate work for the two processes to do this.

### PR DESCRIPTION
**Pull request by dsherret**
With an artificial delay:

https://user-images.githubusercontent.com/1609021/223590348-7c4c6fe1-b10b-4b28-8342-4f1221d60da5.mp4

This is implemented in such a way that it should still allow processes to go through when a file lock wasn't properly cleaned up and the OS hasn't released it yet (but with a 200ms-ish delay).

Closes #18039

### CODE REVIEW

File: cli/util/progress_bar/mod.rs [hunk 1]
```diff
@@ -255,17 +277,25 @@ impl ProgressBar {
   }
 
   pub fn update(&self, msg: &str) -> UpdateGuard {
+    self.update_with_prompt(ProgressMessagePrompt::Download, msg)
+  }
```
Comment (dsherret) [inline cli/util/progress_bar/mod.rs:281]: I could have spent some time making our progress bars more generic, but I didn't think it was worth the time atm.

File: cli/util/fs.rs [hunk 2]
```diff
@@ -471,11 +475,166 @@ pub fn dir_size(path: &Path) -> std::io::Result<u64> {
...
+              // Spawn a blocking task that will continually update a file
+              // signalling the lock is alive. This is a fail safe for when
+              // a file lock is never released. For example, on some operating
+              // systems, if a process does not release the lock (say it's
+              // killed), then the OS may release it at an indeterminate time
+              //
+              // This uses a blocking task because we use a single threaded
+              // runtime and this is time sensitive so we don't want it to update
+              // at the whims of of whatever is occurring on the runtime thread.
+              tokio::task::spawn_blocking({
```
Comment (bartlomieju) [inline cli/util/fs.rs:545]: Cool 👍 I was worried about this scenario, but you handled it nicely.

File: cli/util/progress_bar/mod.rs (hunk 1 above)
Comment (bartlomieju) [reply cli/util/progress_bar/mod.rs:281]: As I side note: we should lazily create `ProgressBar` on `ProcState` - it currently shows up in the flamegraphs for each startup when it's checking if we can actually draw. (even if we never use the progress bar, like when only loading local files)

**Review by bartlomieju — APPROVED** [review]
LGTM

File: cli/util/progress_bar/mod.rs (hunk 1 above)
Comment (dsherret) [reply cli/util/progress_bar/mod.rs:281]: I'll lazily create it here

Comment (dsherret) [reply cli/util/progress_bar/mod.rs:281]: Oh wait, not worth it because `IS_TTY_WITH_CONSOLE_SIZE` is already in a lazy and it would have hit that by now.

**Pull request merged**